### DISCUSSION TITLE
Exemplars: Improve sampling

### DISCUSSION THREAD

**Discussion by josephDunne** [grafana/grafana#41972] [Observability]
**What happened**: plotting a histogram percentiles with exemplars only shows the first exemplar per time bucket

**What you expected to happen**: expected grafana to plot exemplars in the same way that prometheus does - with all exemplars visible

**How to reproduce it (as minimally and precisely as possible)**: create a latency histogram with multiple exemplars. exemplars have a single label trace_id.

**Anything else we need to know?**: Attached is the exemplar query response

**Environment**:
- Grafana version: 8.2.3
- Data source type & version: Prometheus 
- OS Grafana is installed on: Linux
- User OS & Browser: Mac Os & Safari
- Grafana plugins:
- Others:

[query_exemplars.json.zip](https://github.com/grafana/grafana/files/7489006/query_exemplars.json.zip)

![prometheus](https://user-images.githubusercontent.com/1827457/140580286-5bab5951-6dbb-4332-8f16-f7e655d6fc87.png)
![grafana](https://user-images.githubusercontent.com/1827457/140580304-7b3c33ce-65f1-4b00-8047-582c3247a8cd.png)

**Comment by ashharrison90** [comment]
hi o11y friends 👋  

apologies if this doesn't fall under your remit, but i'm not too familiar with exemplars so not sure if this is a bug or working as designed. (or maybe even a problem specific to the timeseries panel!)

**Comment by JokerQueue** [comment]
+1，I have the same problem,  is there any solution?

**Comment by ivanahuckova** [comment]
Hello @josephDunne. Thank you for creating the issue. It is true that Grafana is not showing all exemplars and we are doing sampling on them. This way, the graph is not overflowing when you have thousands of exemplars. We do it here https://github.com/grafana/grafana/blob/main/pkg/tsdb/prometheus/time_series_query.go#L361-L422 and a lot of time this is very useful, but I can see your usecase and why you would like to see them all/not sample them. If it is okay with you, I'll change this to feature request/discussion and add open a discussion about if we should have option to not sample them, or sample them only if there is more than X number of exemplars.

**Reply by josephDunne** [reply]
Hi Ivana, I think its fine to only sample exemplars but the sampling algorithm used is biasing high frequency exemplars. I think its valuable to ensure we can see low frequency exemplars too - so perhaps the problem is when multiple series are used its only sampling from one of the series? i.e. form the 50th percentile in my example but nothing from the 90th percentile

**Reply by ivanahuckova** [reply]
Yeah, that make sense.

**Reply by cstyan** [reply]
Returning 'appropriate' exemplars for `histogram_quantile` queries requires a bit more work if we want to do it in a smart way I think. The resulting series of a `histogram_quantile` query don't contain the same labelsets as the series within the aggregation function, `e2e_latency_bucket` in your case, which is what we use to find 'relevant' exemplars via the exemplar query API.

Overall I think sampling is fine, in most cases returning ALL exemplars is just going to be too noisy to be useful. Sampling in a smarter way might be _nice to have_.

**Reply by josephDunne** [reply]
I would argue that if tools such as Tempo are going to be useful then having working percentile graphs is more than a nice to have. Can Grafana just copy the code prometheus uses?

**Reply by JokerQueue** [reply]
Could we just add a switch for showing ALL exemplars besides finding out a smarter way to sample? A fluctuating series result with some unfluctuating exemplars(high frequency exemplars) is really confused and useless to most of my cases now .

**Reply by JokerQueue** [reply]
I have tried the latest version 8.3.0, the exemplars looks better,  and I realised that showing ALL exemplars is not a good idea.

**Reply by josephDunne** [reply]
yip 8.3.0 seems to have resolved the issue!

**Comment by itsmylife** [comment]
As @grafana/observability-metrics we think this is resolved already. So we are closing the discussion. 
Here is our sampling logic: https://github.com/grafana/grafana/blob/main/pkg/tsdb/prometheus/querydata/exemplar/sampler_stddev.go
Also, we believe that sampling should be provided by Prometheus. 

Even when we think that is resolved the submission encourages conversations and ideas within our development team to continue making improvements, so please keep sharing!

**Comment by cstyan** [comment]
> Also, we believe that sampling should be provided by Prometheus.

It's unlikely that this would happen.
